src/BusinessLayer/SongController/SongController.cpp

```diff
@@ -0,0 +1,97 @@
+#include "SongController.h"
+#include "I_SongPlayerState.h"
+#include "I_SongPlayer.h"
+#include "I_SongUrlSource.h"
+
+#include <QMediaContent>
+#include <QRandomGenerator>
+#include <QStack>
+
+SongController::SongController(I_SongPlayer& songPlayer, I_SongPlayerState& songPlayerState,
+                               I_SongUrlSource& songUrlSource)
+    : songPlayer_(songPlayer)
+    , songPlayerState_(songPlayerState)
+    , songUrls_(songUrlSource.getSongList())
+    , songIndex_(0)
+    , generator_(QRandomGenerator::system())
+    , previousSongs_(new QStack<int>)
+{
+    connect(&songPlayerState_, SIGNAL(playingStateChanged()), this, SLOT(toggleSongPlayingState()));
+    connect(&songPlayerState_, SIGNAL(volumeStateChanged()), this, SLOT(changeVolumeState()));
+    connect(&songPlayer_, SIGNAL(mediaStatusChanged(QMediaPlayer::MediaStatus)), this, SLOT(checkSongEnded(QMediaPlayer::MediaStatus)));
+    loadSong();
+}
+
+SongController::~SongController()
+{
+}
+
+void SongController::playNext()
+{
+    if (songUrls_.size() == 0)
+    {
+        return;
+    }
+
+    //Add the current song to previously played songs
```

Comment:
Remove comment

Reply:
I don't completely agree, but I'll remove the comments.